Add title rewriting

aledeg opened the pull request:
Closes #4335, #4684

Changes proposed in this pull request:

- Rewrite article title upon creation.

How to test the feature manually:

1. Add a new title rewriting rule in the feed configuration page.
2. Load new articles.
3. Check that the titles are following the rewrite rule.
4. Enjoy!

Pull request checklist:

- [x] clear commit messages
- [x] code manually tested
- [x] unit tests written (optional if too hard)
- [x] documentation updated

[Additional information can be found in the documentation](https://github.com/FreshRSS/FreshRSS/tree/edge/docs/en/developers/04_Pull_requests.md).

The title rewriting feature allows title rewriting upon article creation in the database. Once the rewriting is done, it cannot be undone.

At the moment, there is 2 variables and 3 filters available in the simple templating system.

aledeg commented:
There is known limitations but I can address that later if it's blocking.
There is only 2 variables and 3 filters supported but it's easy to expand the list if needed.

nbgibson commented:
Ok, that took a bit more doing than I care to admit (entirely on my end), but I got things up and running. I do see the title rewrite field as expected, but I'm a bit stumped on how the rules work at first glance. Can you point me at some documentation on how they work or a quick example? Apologies if this is a known quantity, I've got baby brain in a bad way right now.

nbgibson commented:
Ok, that took a bit more doing than I care to admit (entirely on my end), but I got things up and running. I do see the title rewrite field as expected, but I'm a bit stumped on how the rules work at first glance. Can you point me at some documentation on how they work or a quick example? Apologies if this is a known quantity, I've got baby brain in a bad way right now.

aledeg commented:
You can check the documentation in that PR.
If you have hard time to find it. Here is a link on my repo https://github.com/aledeg/FreshRSS/blob/feed/title-rewriting/docs/en/users/04_Subscriptions.md#title-rewriting

nbgibson commented:
Perfect, thank you. I've applied some rules and am awaiting to see how it goes. We should have some results by the morning.

nbgibson commented:
First glance through things appear to be working as intended. I did encounter what appeared to be a bug where entering a filter would cause titles to appear as just the article URLs, but I'm working on replicating it. It could be the LAMP stack I have running this being weird/dumb.

nbgibson reviewed (approved):
I've spent some time playing around with this feature set and everything appears to be working as expected. I did see a one off bug where titles would be rendered as their source URLs, but I've been unable to replicate it and I suspect it was on the side of my specific setup and not the fault of the code.

aledeg commented:
If you happen to find out why, let me know. I know there are some issues with some special characters. But we might address that later.

ColonelMoutarde commented on app/Handlers/TitleRewritingHandler.php:
```diff
@@ -0,0 +1,272 @@
+				2
+			],
+		],
+	];
```
Pleas add the visibility of the constant like "public" in this case

aledeg replied:
Done

ColonelMoutarde commented on app/Handlers/TitleRewritingHandler.php:
```diff
@@ -0,0 +1,272 @@
+		$this->cleanRule();
+	}
+
+	private function parseRule(string $rule) {
```
Please, split this method for better readability and maintainability

aledeg replied:
I am not sure if it will make this more readable. I'll try something.

ColonelMoutarde commented on app/Models/Feed.php:
```diff
@@ -200,6 +200,13 @@ public function attributes($key = '') {
 		}
 	}
 
+	public function titleRewriteHandler() {
```
```suggestion
	public function titleRewriteHandler(): FreshRSS_TitleRewriting_Handler {
```

aledeg replied:
I can not do that since this method could return void. With PHP7.2, I can not type-hint the 2 values.

ColonelMoutarde commented on app/Handlers/TitleRewritingHandler.php:
```diff
@@ -0,0 +1,272 @@
+<?php
+
```
suggestion: 
when you add a new class, it is wise to add this at the beginning of the file so that php does not transform the types in an uncontrollable way.

```suggestion

declare(strict_types=1);

```

aledeg replied:
Done

ColonelMoutarde commented on app/Models/Feed.php:
```diff
@@ -200,6 +200,13 @@ public function attributes($key = '') {
 		}
 	}
 
+	public function titleRewriteHandler() {
```
you can add a php like that:

```suggestion
        /**
        * @return toto|bool
        */
	public function titleRewriteHandler() {
```

aledeg replied:
Done

ColonelMoutarde commented on tests/app/Handlers/TitleRewritingHandlerTest.php:
```diff
@@ -0,0 +1,470 @@
+<?php
+
```
```suggestion

declare(strict_types=1);

```

aledeg replied:
Done

ColonelMoutarde commented on tests/app/Handlers/TitleRewritingHandlerTest.php:
```diff
@@ -0,0 +1,470 @@
+	/**
+	 * @dataProvider provideValidTemplate
+	 */
+	public function test__construct_whenValidTemplate_storeRule($template, $rule) {
```
```suggestion
	public function test__construct_whenValidTemplate_storeRule($template, $rule): void {
```

aledeg replied:
Done

Alkarex commented:
I had a bit forgotten about that PR, sorry @aledeg :-/
Back then, I postponed it because although the functionality is nice, it looks like the implementation is really huge (relatively to the task).
I have to admit I have not given it much thoughts, but I am wondering whether a few lines of code just allowing a regex-like replace wouldn't do the job, while also being simpler, more powerful, and more standard.

nbgibson commented:
It would likely be all those things, but it would also be a fairly significant ask on the average end user to learn and implement a Regex filter to achieve the desired result. It would also be a deviation from the existing article filter rule implementation methods.

Unless you mean on the back end and not end user facing. If so then just ignore me.

Alkarex commented:
> Unless you mean on the back end and not end user facing.

My main concern is indeed the (backend) implementation. But on the frontend, it is quite also usual to have a toggle to allow the more expressive regex syntax. But again, I have to admit that I have not yet spent much time thinking about the current case.

aledeg commented:
> I had a bit forgotten about that PR, sorry @aledeg :-/ Back then, I postponed it because although the functionality is nice, it looks like the implementation is really huge (relatively to the task). I have to admit I have not given it much thoughts, but I am wondering whether a few lines of code just allowing a regex-like replace wouldn't do the job, while also being simpler, more powerful, and more standard.

@Alkarex don't worry about the delay.

I know the implementation of a parser is quite a lot of code. But I figured that it might be extended in an easier way compared to regex.
I figured that if we make it powerful enough, we could reuse it somewhere else for various purposes. It easy to add modifiers (bold, emphasis, underline, ...) and to improve the parsing.
It could be use as a templating engine to modify article content, or even the application itself. We can also make it extendable to be used in extensions.

If you think it would be better to go some other directions, feel free to close that PR.

As @nbgibson mentioned, I think that offering regex can be tricky for a lot of users, even simple one.

chenyucheng0503 commented:
Can we add the author name to rewrite title?

aledeg commented:
Probably not as-is. But with some tweaking, it might be possible. I have to make time to work on that.

aledeg committed:
Add title rewriting

The title rewriting feature allows title rewriting upon article creation in the
database. Once the rewriting is done, it cannot be undone.

At the moment, there is 2 variables and 3 filters available in the simple
templating system.
